[WIP] Revamped octane tutorial (#1002)
ember-learn/guides-source: guides/release/tutorial/14.md

So far, all the components we have written are purely *[presentational](TODO: link to presentational)* &mdash; they are simply reusable snippets of markup. That's pretty cool! But in Ember, components can do so much more.

Sometimes, you want to associate some *[behavior](TODO: link to behavior)* with your components so that they can do more interesting things. For example, `<LinkTo>` can respond to clicks by changing the URL and navigating us to a different page.

> Here, we are going to do just that! We are going to implementing the "View Larger" and "View Smaller" functionality. Specifically, we want to give our users the ability to view a larger version of the property's image by clicking on it, and when they are done, they can click on it again to go back to the smaller version.

Reviewer: ```suggestion
Here, we are going to do just that! We are going to implement the "View Larger" and "View Smaller" functionality, which will allow our users to click on a property's image to view a larger version, and click on it again to return to the smaller version.
```

> In other words, we want a way to _toggle_ the image between one of the two *[states](TODO: link to states)*. In order to do that, we need a way for the component to store two potential states, and to also be aware which of the two possible states it is currently in.

Reviewer: ```suggestion
In other words, we want a way to _toggle_ the image between one of the two *[states](TODO: link to states)*. In order to do that, we need a way for the component to store two possible states, and to be aware of which state it is currently in.
```

> Ember allows us to optionally associate JavaScript code with a component for exactly this purpose. We can generate a JavaScript file for our `<Rental::Image>` component by running the `component-class` generator:

Reviewer: ```suggestion
Ember optionally allows us to associate JavaScript code with a component for exactly this purpose. We can generate a JavaScript file for our `<Rental::Image>` component by running the `component-class` generator:
```

```
ember generate component-class rental/image
```

> This generated a JavaScript file with the same name as our component's template at `app/components/rentals/image.js`. It contains a *[JavaScript class](TODO: link to JavaScript class)*, subclassing from `@glimmer/component`.

Reviewer: ```suggestion
This generated a JavaScript file with the same name as our component's template at `app/components/rentals/image.js`. It contains a *[JavaScript class](TODO: link to JavaScript class)* inheriting from `@glimmer/component`.
```

<div class="cta">
  <div class="cta-note">
    <div class="cta-note-body">
      <div class="cta-note-heading">Zoey says...</div>
      <div class="cta-note-message">
        <p>`@glimmer/component`, or *[Glimmer component](TODO: link to Glimmer component)*, is one of the several component classes available to use. Component classes are a great starting point whenever you want to add behavior to your components. In this tutorial, we will be using Glimmer components exclusively.</p>

>         <p>In general, Glimmer components should be used whenever possible. However, you may also see `@ember/components`, or *[classic components](TODO: link to classic components)*, being used in older apps. You can tell them apart from looking at their import path, which is helpful for looking up the respective documentation, as they have different and incompatible APIs.</p>

Reviewer: ```suggestion
        <p>In general, Glimmer components should be used whenever possible. However, you may also see `@ember/components`, or *[classic components](TODO: link to classic components)* used in older apps. You can tell them apart by looking at their import path (which is helpful for looking up the respective documentation, as they have different and incompatible APIs).</p>
```

      </div>
    </div>
    <img src="/images/mascots/zoey.png" role="presentation" alt="Ember Mascot">
  </div>
</div>

> Ember will create an *[instance](TODO: link to instance)* of the class whenever our component is invoked. We can use that to store our state:

Reviewer: ```suggestion
Ember will create an *[instance](TODO: link to instance)* of the class whenever our component is invoked. We can use that instance to store our state:
```

```js {data-filename="app/components/rental/image.js" data-diff="+4,+5,+6"}
import Component from '@glimmer/component';

export default class RentalImageComponent extends Component {
  constructor(...args) {
    super(...args);
    this.isLarge = false;
  }
}
```

> Here, in the *[component's constructor](todo://link to component's constructor)*, we *[initialized](TODO: link to initialized)* the *[instance variable](TODO: link to instance variable)* `this.isLarge` to the value `false`, since this is the default state that we want for our component.

Reviewer: ```suggestion
Here, in the *[component's constructor](todo://link to component's constructor)*, we *[initialized](TODO: link to initialized)* the *[instance variable](TODO: link to instance variable)* `this.isLarge` with the value `false`, since this is the default state that we want for our component.
```

Let's update our template to use this state we just added:

```handlebars {data-filename="app/components/rental/image.hbs"}
{{#if this.isLarge}}
  <div class="image large">
    <img ...attributes>
    <small>View Smaller</small>
  </div>
{{else}}
  <div class="image">
    <img ...attributes>
    <small>View Larger</small>
  </div>
{{/if}}
```

> The `{{#if ...}}...{{else}}...{{/if}}` *[conditional](TODO: link to conditional)* syntax allows us to render different content. We also have access to the component's instance variables from the template. Combining these two features, we can render either the small or the large version of the image accordingly.

Reviewer: ```suggestion
In the template, we have access to the component's instance variables. The `{{#if ...}}...{{else}}...{{/if}}` *[conditional](TODO: link to conditional)* syntax allows us to render different content depending on the result of the _predicate_ (in this case, the instance variable`this.isLarge`). Combining these two features, we can render either the small or the large version of the image accordingly.
```

> We can verify this is working by temporarily changing the initial value in our JavaScript file. If we change `app/components/rental/image.js` to initialize `this.isLarge = true;` in the constructor, we should see the large version of the property image in the browser. Cool!

Reviewer: ```suggestion
We can verify this works by temporarily changing the initial value in our JavaScript file. If we change `app/components/rental/image.js` to initialize `this.isLarge = true;` in the constructor, we should see the large version of the property image in the browser. Cool!
```

<!-- TODO: screenshot? -->

Once we've tested this out, we can change `this.isLarge` back to `false`.

> Since this pattern of initializing instance variables from the constructor is pretty common, there happens to be a much more concise syntax for it:

Reviewer: ```suggestion
Since this pattern of initializing instance variables in the constructor is pretty common, there happens to be a much more concise syntax for it:
```

```js {data-filename="app/components/rental/image.js" data-diff="+4,-5,-6,-7,-8"}
import Component from '@glimmer/component';

export default class RentalImageComponent extends Component {
  isLarge = false;
  constructor() {
    super(...arguments);

>     this.isLarge = false;

Reviewer: Should this line still be here? May want to say something like "we can declare default values for instance variables outside the constructor"
